File: core/src/com/bombbird/atcsim/screens/GameScreen.java
package com.bombbird.atcsim.screens;

import com.badlogic.gdx.*;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.input.GestureDetector;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.math.Vector3;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.viewport.Viewport;
import com.bombbird.atcsim.AtcSim;
import com.bombbird.atcsim.entities.Airport;
import com.bombbird.atcsim.entities.ILS;
import com.bombbird.atcsim.entities.RangeCircle;
import com.bombbird.atcsim.entities.Waypoint;
import com.bombbird.atcsim.entities.aircrafts.Aircraft;
import com.bombbird.atcsim.entities.restrictions.Obstacle;
import com.bombbird.atcsim.entities.restrictions.RestrictedArea;
import com.bombbird.atcsim.screens.Ui.Ui;
import com.bombbird.atcsim.utilities.Fonts;

import java.util.HashMap;

public class GameScreen implements Screen, GestureDetector.GestureListener, InputProcessor {
    //Init game (set in constructor)
    private final AtcSim game;
    public static Stage stage;
    private boolean aircraftLoaded;
    public boolean loading;
    public String loadingPercent;
    private float loadedTime = 0;

    //Set input processors
    public InputMultiplexer inputMultiplexer = new InputMultiplexer();
    public InputProcessor inputProcessor1 = this;
    public InputProcessor inputProcessor2;
    public InputProcessor inputProcessor3;
    public GestureDetector gd = new GestureDetector(40, 0.2f,1.1f, 0.15f, this);

    //Pinch zoom constants
    private float lastInitialDist = 0;
    private float lastScale = 1;

    //Double tap animation variables
    private boolean zooming = false;
    private boolean zoomedIn = false;

    //Create new camera
    public OrthographicCamera camera;
    public Viewport viewport;
    private float lastZoom = 1;

    //Create 2nd camera for UI
    public static Ui ui;
    public static Stage uiStage;
    public OrthographicCamera uiCam;
    public Viewport uiViewport;

    //Create texture stuff
    public Skin skin;
    public static ShapeRenderer shapeRenderer;

    //Create range circles
    private RangeCircle[] rangeCircles;

    //Create obstacle resources
    public Array<Obstacle> obsArray;
    public Array<RestrictedArea> restArray;

    //Create airports + wind data
    public static HashMap<String, Airport> airports;

    //Array of planes
    public static HashMap<String, Aircraft> aircrafts;

    //Create waypoints
    public static HashMap<String, Waypoint> waypoints;

    public GameScreen(final AtcSim game) {
        this.game = game;
        shapeRenderer = new ShapeRenderer();

        //Initiate range circles
        rangeCircles = new RangeCircle[3];

        //Initiate airports
        airports = new HashMap<String, Airport>();

        loading = false;
        aircraftLoaded = false;
        loadingPercent = "0%";
    }

    public void loadRange() {
        //Load radar screen range circles
        rangeCircles[0] = new RangeCircle(10, -255);
        rangeCircles[1] = new RangeCircle(30, -900);
        rangeCircles[2] = new RangeCircle(50, -1548);
        for (RangeCircle circle: rangeCircles) {
            stage.addActor(circle);
        }
    }

    private void handleInput(float dt) {
        //Handles input from keyboard, mouse, moderates them
        float ZOOMCONSTANT = 0.6f;
        float SCROLLCONSTANT = 150;
        if (Gdx.input.isKeyPressed(Input.Keys.A)) {
            //Zoom in
            camera.zoom += ZOOMCONSTANT * dt;
        }
        if (Gdx.input.isKeyPressed(Input.Keys.Q)) {
            //Zoom out
            camera.zoom -= ZOOMCONSTANT * dt;
        }
        if (Gdx.input.isKeyPressed(Input.Keys.LEFT)) {
            //Move left
            camera.translate(-SCROLLCONSTANT / camera.zoom * dt, 0, 0);
        }
        if (Gdx.input.isKeyPressed(Input.Keys.RIGHT)) {
            //Move right
            camera.translate(SCROLLCONSTANT / camera.zoom * dt, 0, 0);
        }
        if (Gdx.input.isKeyPressed(Input.Keys.DOWN)) {
            //Move down
            camera.translate(0, -SCROLLCONSTANT / camera.zoom * dt, 0);
        }
        if (Gdx.input.isKeyPressed(Input.Keys.UP)) {
            //Move up
            camera.translate(0, SCROLLCONSTANT / camera.zoom * dt, 0);
        }

        if (zooming) {
            if (zoomedIn) {
                camera.zoom += 0.05f;
            } else {
                camera.zoom -= 0.05f;
            }
        }

        //Make sure user doesn't zoom in too much or zoom out of bounds
        if (camera.zoom < 0.3f) {
            camera.zoom = 0.3f;
            zooming = false;
            zoomedIn = true;
        } else if (Gdx.app.getType() == Application.ApplicationType.Android && camera.zoom > 0.6f && !loading) {
            camera.zoom = 0.6f;
            zooming = false;
            zoomedIn = false;
        } else if (camera.zoom > 1.0f) {
            camera.zoom = 1.0f;
            zooming = false;
            zoomedIn = false;
        }
        camera.translate(-990 * (camera.zoom - lastZoom), 0); //Ensure camera zooms into the current center
        lastZoom = camera.zoom;

        //Setting new boundaries for camera position after zooming
        float effectiveViewportWidth = (camera.viewportWidth - ui.getPaneWidth()) * camera.zoom; //Take width of pane into account
        float xDeviation = -(effectiveViewportWidth - camera.viewportWidth) / 2f;
        float effectiveViewportHeight = camera.viewportHeight * camera.zoom;
        float yDeviation = -(effectiveViewportHeight - camera.viewportHeight) / 2f;
        float xOffset = camera.zoom * 990; //Since I shifted camera to the left by 990 px
        float leftLimit = effectiveViewportWidth / 2f - xOffset;
        float rightLimit = leftLimit + 2 * xDeviation;
        float downLimit = effectiveViewportHeight / 2f;
        float upLimit = downLimit + 2 * yDeviation;

        //Prevent camera from going out of boundary
        if (camera.position.x < leftLimit) {
            camera.position.x = leftLimit;
        } else if (camera.position.x > rightLimit) {
            camera.position.x = rightLimit;
        }
        if (camera.position.y < downLimit) {
            camera.position.y = downLimit;
        } else if (camera.position.y > upLimit) {
            camera.position.y = upLimit;
        }
    }

    public void newAircraft() {
        //Creates new aircraft; overridden in radarscreen subclass
    }

    @Override
    public void show() {
        //Implements show method of screen; overridden in radarscreen class
    }

    public void renderShape() {
        //Renders shapes for aircraft trajectory line, obstacle and restricted areas; overridden in radarscreen class
    }

    @Override
    public void render(float delta) {
        //Main rendering method for rendering to spritebatch
        if (!loading && !aircraftLoaded) {
            //Load the initial aircrafts if METAR has finished loading but aircrafts not loaded yet
            newAircraft();
            aircraftLoaded = true;
        }

        //Clear screen
        Gdx.gl.glClearColor(0, 0, 0, 0);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);

        //Test for input, update camera
        if (!loading) {
            handleInput(delta);
        }
        camera.update();

        //Set rendering for stage camera
        game.batch.setProjectionMatrix(camera.combined);
        shapeRenderer.setProjectionMatrix(camera.combined);

        //Update stage
        stage.act(delta);

        //Render each of the range circles, obstacles using shaperenderer
        if (!loading) {
            stage.getViewport().apply();
            //Render shapes only if METAR has finished loading
            shapeRenderer.begin(ShapeRenderer.ShapeType.Line);
            for (RangeCircle rangeCircle : rangeCircles) {
                rangeCircle.renderShape();
            }
            renderShape();
            shapeRenderer.end();
        }

        //Draw to the spritebatch
        game.batch.begin();
        String loadingText = "Loading.   ";
        if (loading) {
            //Write loading text if loading
            loadedTime += Gdx.graphics.getDeltaTime();
            if (loadedTime > 1.5) {
                loadedTime = 0;
                loadingText = "Loading.   ";
            } else if (loadedTime > 1) {
                loadingText = "Loading... ";
            } else if (loadedTime > 0.5) {
                loadingText = "Loading..  ";
            }
            Fonts.defaultFont20.draw(game.batch, loadingText + loadingPercent, 1560, 1550);
        } else {
            stage.draw();
        }
        game.batch.end();

        //Draw the UI overlay
        uiCam.update();
        if (!loading) {
            game.batch.setProjectionMatrix(uiCam.combined);
            uiStage.act();
            game.batch.begin();
            uiStage.getViewport().apply();
            uiStage.draw();
            game.batch.end();
        }
    }

    @Override
    public void resize(int width, int height) {
        //Implements resize method of screen, adjusts camera & viewport properties after resize for better UI
        AtcSim.WIDTH = width;
        AtcSim.HEIGHT = height;

        ui.updatePaneWidth();

        viewport.update(width, height, false);
        stage.getViewport().update(width, height, false);
        float xOffset = camera.zoom * 990;
        camera.position.set(camera.viewportWidth / 2 - xOffset, camera.viewportHeight / 2, 0);

        uiViewport.update(width, height, true);
        uiStage.getViewport().update(width, height, true);
        uiCam.position.set(uiCam.viewportWidth / 2f, uiCam.viewportHeight / 2f, 0);
        if (Gdx.app.getType() == Application.ApplicationType.Desktop) {
            boolean resizeAgain = false;
            if (width < 960) {
                width = 960;
                resizeAgain = true;
            }
            if (height < 540) {
                height = 540;
                resizeAgain = true;
            }
            if (resizeAgain) {
                resize(width, height);
            }
        }
    }

    @Override
    public void pause() {
        //Implements pause method of screen
    }

    @Override
    public void resume() {
        //Implements pause method of screen
    }

    @Override
    public void hide() {
        //Implements hide method of screen
        dispose();
    }

    @Override
    public void dispose() {
        //Implements dispose method of screen, disposes resources after they're no longer needed
        shapeRenderer.dispose();
        stage.clear();
        stage.dispose();
        skin.dispose();
    }

    @Override
    public boolean touchDown(float x, float y, int pointer, int button) {
        //Implements touchdown method of gesturelistener
        return false;
    }

    @Override
    public boolean tap(float x, float y, int count, int button) {
        //Implements tap method of gesturelistener, tests for tap and double tap
        RadarScreen.setSelectedAircraft(null);
        if (count == 2 && !loading) {
            zooming = true;
            return true;
        }
        return false;
    }

    @Override
    public boolean longPress(float x, float y) {
        //Implements longpress method of gesturelistener
        return false;
    }

    @Override
    public boolean fling(float velocityX, float velocityY, int button) {
        //Implements fling method of gesturelistener
        return false;
    }

    @Override
    public boolean pan(float x, float y, float deltaX, float deltaY) {
        //Implements pan method of gesturelistener, tests for panning to shift radar screen around
        if (loading) {
            return false;
        }
        float screenHeightRatio = 3240f / Gdx.graphics.getHeight();
        float screenWidthRatio = 5760f / Gdx.graphics.getWidth();
        float screenRatio = (screenHeightRatio > screenWidthRatio) ? screenHeightRatio : screenWidthRatio;
        camera.translate(-deltaX * camera.zoom * screenRatio, deltaY * camera.zoom * screenRatio);
        return false;
    }

    @Override
    public boolean panStop(float x, float y, int pointer, int button) {
        //Implements panstop method of gesturelistener
        return false;
    }

    @Override
    public boolean zoom(float initialDistance, float distance) {
        //Implements zoom method of gesturelistener, tests for pinch zooming to adjust radar screen zoom level
        if (loading) {
            return true;
        }
        if (initialDistance != lastInitialDist) {
            //New zoom
            lastInitialDist = initialDistance;
            lastScale = camera.zoom;
        }
        float ratio = lastInitialDist / distance;
        camera.zoom = lastScale * ratio;
        camera.update();
        return true;
    }

    @Override
    public boolean pinch(Vector2 initialPointer1, Vector2 initialPointer2, Vector2 pointer1, Vector2 pointer2) {
        //Implements pinch method of gesturelistener
        return false;
    }

    @Override
    public void pinchStop() {
        //Implements pinchstop method of gesturelistener
    }

    @Override
    public boolean keyDown(int keycode) {
        //Implements keydown method of inputlistener
        return false;
    }

    @Override
    public boolean keyUp(int keycode) {
        //Implements keyup method of inputlistener
        return false;
    }

    @Override
    public boolean keyTyped(char character) {
        //Implements keytyped method of inputlistener
        return false;
    }

    @Override
    public boolean touchDown(int screenX, int screenY, int pointer, int button) {
        //Implements touchdown method of inputlistener
        return false;
    }

    @Override
    public boolean touchUp(int screenX, int screenY, int pointer, int button) {
        //Implements touchup method of inputlistener
        return false;
    }

    @Override
    public boolean touchDragged(int screenX, int screenY, int pointer) {
        //Implements touchdragged method of inputlistener
        return false;
    }

    @Override
    public boolean mouseMoved(int screenX, int screenY) {
        //Implements mousemoved method of inputlistener
        return false;
    }

    @Override
    public boolean scrolled(int amount) {
        //Implements scrolled method of inputlistener, tests for scrolling to adjust radar screen zoom levels
        if (!loading) {
            camera.zoom += amount * 0.042f;
        }
        return true;
    }
}

File: core/src/com/bombbird/terminalcontrol/screens/gamescreen/GameScreen.kt
package com.bombbird.terminalcontrol.screens.gamescreen

import com.badlogic.gdx.*
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import com.badlogic.gdx.input.GestureDetector
import com.badlogic.gdx.input.GestureDetector.GestureListener
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.math.Vector3
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.viewport.Viewport
import com.bombbird.terminalcontrol.TerminalControl
import com.bombbird.terminalcontrol.entities.RangeCircle
import com.bombbird.terminalcontrol.entities.aircrafts.Aircraft
import com.bombbird.terminalcontrol.entities.airports.Airport
import com.bombbird.terminalcontrol.entities.obstacles.Obstacle
import com.bombbird.terminalcontrol.entities.waypoints.Waypoint
import com.bombbird.terminalcontrol.screens.MainMenuScreen
import com.bombbird.terminalcontrol.screens.PauseScreen
import com.bombbird.terminalcontrol.sounds.SoundManager
import com.bombbird.terminalcontrol.ui.DataTag
import com.bombbird.terminalcontrol.ui.RandomTip.loadTips
import com.bombbird.terminalcontrol.ui.RandomTip.randomTip
import com.bombbird.terminalcontrol.ui.RandomTip.tipsLoaded
import com.bombbird.terminalcontrol.ui.RequestFlasher
import com.bombbird.terminalcontrol.ui.Ui
import com.bombbird.terminalcontrol.utilities.errors.ErrorHandler
import com.bombbird.terminalcontrol.utilities.Fonts
import com.bombbird.terminalcontrol.utilities.SafeStage
import kotlin.collections.HashMap

open class GameScreen(val game: TerminalControl) : Screen, GestureListener, InputProcessor {
    lateinit var stage: SafeStage
    lateinit var labelStage: SafeStage
    var uiLoaded = false
    var metarLoading = false
    var loadingTime = 0f
    private var loadedTime = 0f
    private lateinit var loadingLabel: Label
    private lateinit var tipLabel: Label

    //Flag whether to quit the tutorial on next loop
    var tutorialQuit = false

    //Play time timer
    var playTime: Float

    //Set input processors
    var inputMultiplexer = InputMultiplexer()
    lateinit var gd: GestureDetector

    //Pinch zoom constants
    private var lastInitialDist = 0f
    private var lastScale = 1f

    //Double tap animation variables
    private var zooming = false
    private var zoomedIn = false

    //Create new camera
    lateinit var camera: OrthographicCamera
    lateinit var viewport: Viewport
    private var lastZoom = 1f

    //Create 2nd camera for UI
    lateinit var ui: Ui
    lateinit var uiStage: SafeStage
    lateinit var uiCam: OrthographicCamera
    lateinit var uiViewport: Viewport

    //Create texture stuff
    val shapeRenderer = ShapeRenderer()

    //Flashes rectangles to alert of user of aircraft with request if aircraft not within view
    lateinit var requestFlasher: RequestFlasher

    //Sounds
    var soundManager = SoundManager()

    //Create range circles
    private val rangeCircles: Array<RangeCircle> = Array()
    var rangeCircleDist = 0

    //Create obstacle resources
    var obsArray = Array<Obstacle>()

    //Create airports
    val airports = HashMap<String, Airport>()

    //HashMap of planes
    val aircrafts = HashMap<String, Aircraft>()

    //Create waypoints
    var waypoints: HashMap<String, Waypoint> = HashMap()
    var running = false
    var speed = 1

    //Stores variables for right click drag measuring of distance
    var dragging = false
    var distMode = false //For android only
    var firstPoint = Vector2()
    var secondPoint = Vector2()

    init {
        //Initiate range circles
        loadLabels()
        playTime = 0f
    }

    /** Load loading, tips labels  */
    private fun loadLabels() {
        val labelStyle = LabelStyle()
        labelStyle.font = Fonts.defaultFont20
        labelStyle.fontColor = Color.WHITE
        loadingLabel = Label("", labelStyle)
        val labelStyle1 = LabelStyle()
        labelStyle1.font = Fonts.defaultFont16
        labelStyle1.fontColor = Color.WHITE
        tipLabel = Label("", labelStyle1)
    }

    /** Handles input from keyboard, mouse, moderates them  */
    private fun handleInput(dt: Float) {
        val zoomConstant = 0.6f
        val scrollConstant = 150f
        if (Gdx.input.isKeyPressed(Input.Keys.A)) {
            //Zoom in
            camera.zoom += zoomConstant * dt
        }
        if (Gdx.input.isKeyPressed(Input.Keys.Q)) {
            //Zoom out
            camera.zoom -= zoomConstant * dt
        }
        if (Gdx.input.isKeyPressed(Input.Keys.LEFT)) {
            //Move left
            camera.translate(-scrollConstant / camera.zoom * dt, 0f, 0f)
        }
        if (Gdx.input.isKeyPressed(Input.Keys.RIGHT)) {
            //Move right
            camera.translate(scrollConstant / camera.zoom * dt, 0f, 0f)
        }
        if (Gdx.input.isKeyPressed(Input.Keys.DOWN)) {
            //Move down
            camera.translate(0f, -scrollConstant / camera.zoom * dt, 0f)
        }
        if (Gdx.input.isKeyPressed(Input.Keys.UP)) {
            //Move up
            camera.translate(0f, scrollConstant / camera.zoom * dt, 0f)
        }
        if (zooming) {
            if (zoomedIn) {
                camera.zoom += 0.05f
            } else {
                camera.zoom -= 0.05f
            }
        }
        moderateZoom()
        moderateCamPos()

        if (Gdx.input.isButtonPressed(Input.Buttons.RIGHT) && Gdx.app.type == Application.ApplicationType.Desktop) {
            if (!dragging) {
                firstPoint = getWorldCoordFromScreenCoord(Gdx.input.x.toFloat(), Gdx.input.y.toFloat())
            }
            secondPoint = getWorldCoordFromScreenCoord(Gdx.input.x.toFloat(), Gdx.input.y.toFloat())
            dragging = true
        } else if (Gdx.input.isTouched(0) && Gdx.input.isTouched(1) && Gdx.app.type == Application.ApplicationType.Android && distMode) {
            dragging = true
            firstPoint = getWorldCoordFromScreenCoord(Gdx.input.getX(0).toFloat(), Gdx.input.getY(0).toFloat())
            secondPoint = getWorldCoordFromScreenCoord(Gdx.input.getX(1).toFloat(), Gdx.input.getY(1).toFloat())
        } else {
            dragging = false
            firstPoint.setZero()
            secondPoint.setZero()
        }
    }

    /** Prevents user from over or under zooming  */
    private fun moderateZoom() {
        val minZoom = if (TerminalControl.increaseZoom) 0.2f else 0.3f
        val maxZoomAndroid = 0.6f
        val maxZoomDesktop = 1.0f
        if (camera.zoom < minZoom) {
            camera.zoom = minZoom
            zooming = false
            zoomedIn = true
        } else if (Gdx.app.type == Application.ApplicationType.Android && camera.zoom > maxZoomAndroid && finishedLoading) {
            camera.zoom = maxZoomAndroid
            zooming = false
            zoomedIn = false
        } else if (camera.zoom > maxZoomDesktop) {
            camera.zoom = maxZoomDesktop
            zooming = false
            zoomedIn = false
        }
        camera.translate(-990 * (camera.zoom - lastZoom), 0f) //Ensure camera zooms into the current center
        lastZoom = camera.zoom
    }

    /** Moderates the position of the camera to prevent it from going out of boundary  */
    private fun moderateCamPos() {
        //Setting new boundaries for camera position after zooming
        val effectiveViewportWidth = (camera.viewportWidth - ui.paneWidth) * camera.zoom //Take width of pane into account
        val xDeviation = -(effectiveViewportWidth - camera.viewportWidth) / 2f
        val effectiveViewportHeight = camera.viewportHeight * camera.zoom
        val yDeviation = -(effectiveViewportHeight - camera.viewportHeight) / 2f
        val xOffset = camera.zoom * 990 //Since I shifted camera to the left by 990 px
        val leftLimit = effectiveViewportWidth / 2f - xOffset
        val rightLimit = leftLimit + 2 * xDeviation
        val downLimit = effectiveViewportHeight / 2f
        val upLimit = downLimit + 2 * yDeviation

        //Prevent camera from going out of boundary
        if (camera.position.x < leftLimit) {
            camera.position.x = leftLimit
        } else if (camera.position.x > rightLimit) {
            camera.position.x = rightLimit
        }
        if (camera.position.y < downLimit) {
            camera.position.y = downLimit
        } else if (camera.position.y > upLimit) {
            camera.position.y = upLimit
        }
    }

    /** Implements show method of screen; overridden in radarScreen class  */
    override fun show() {
        //No default implementation
    }

    /** Renders shapes for aircraft trajectory line, obstacle and restricted areas, APW, STCAS and others; overridden in radarScreen class  */
    open fun renderShape() {
        //No default implementation
    }

    /** Load radar screen range circles  */
    fun loadRange() {
        for (circle in rangeCircles) {
            //Clear any existing circles
            circle.remove()
        }
        rangeCircles.clear()
        if (rangeCircleDist == 0) return
        var i = rangeCircleDist
        while (i < 50) {
            val circle = RangeCircle(i)
            rangeCircles.add(circle)
            stage.addActor(circle)
            i += rangeCircleDist
        }
    }

    /** Renders the range circles  */
    fun renderRangeCircles() {
        for (rangeCircle in rangeCircles) {
            rangeCircle.renderShape()
        }
    }

    /** Updates timer specifically for tutorial  */
    open fun updateTutorial() {
        //No default implementation
    }

    /** Main update method, overridden in radarScreen  */
    open fun update() {
        //No default implementation
    }

    /** Main rendering method for rendering to spriteBatch  */
    override fun render(delta: Float) {
        try {
            //Clear screen
            Gdx.gl.glClearColor(0f, 0f, 0f, 0f)
            Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT or GL20.GL_DEPTH_BUFFER_BIT or if (Gdx.graphics.bufferFormat.coverageSampling) GL20.GL_COVERAGE_BUFFER_BIT_NV else 0)
            if (running) {
                //Test for input, update camera
                if (finishedLoading) {
                    handleInput(delta)
                }
                camera.update()

                //Set rendering for stage camera
                game.batch.projectionMatrix = camera.combined
                shapeRenderer.projectionMatrix = camera.combined

                //Update stage
                for (i in 0 until speed) {
                    if (checkTutorialPaused()) break
                    stage.act(delta)
                    labelStage.act(delta)
                }

                //Render each of the range circles, obstacles using shaperenderer, update loop
                if (finishedLoading) {
                    stage.viewport.apply()
                    for (i in 0 until speed) {
                        updateTutorial() //Tutorial timer is updated even if tutorial is paused
                        if (!checkTutorialPaused()) update()
                    }
                    //Update playtime counter
                    playTime += delta
                    //Render shapes only if METAR has finished loading
                    shapeRenderer.begin(ShapeRenderer.ShapeType.Filled)
                    renderShape()
                    shapeRenderer.end()
                }

                //Draw to the spritebatch
                game.batch.projectionMatrix = labelStage.camera.combined
                val liveWeather = (this as RadarScreen).weatherSel == RadarScreen.Weather.LIVE && !this.tutorial
                if (!finishedLoading) {
                    //Write loading text if loading
                    loadingTime += delta
                    loadedTime += delta
                    val loadingText = when {
                        loadedTime > 1.5 -> {
                            loadedTime = 0f
                            if (liveWeather) "Loading live weather.   " else "Loading.   "
                        }
                        loadedTime > 1 -> if (liveWeather) "Loading live weather... " else "Loading... "
                        loadedTime > 0.5 -> if (liveWeather) "Loading live weather..  " else "Loading..  "
                        else -> if (liveWeather) "Loading live weather.   " else "Loading.   "
                    }
                    loadingLabel.setText(loadingText)
                    loadingLabel.setPosition(1920 - loadingLabel.prefWidth / 2, 1550f)
                    game.batch.begin()
                    loadingLabel.draw(game.batch, 1f)
                    if (!tipsLoaded()) loadTips()
                    try {
                        if ("" == tipLabel.text.toString()) {
                            tipLabel.setText(randomTip())
                            tipLabel.setPosition(1920 - tipLabel.prefWidth / 2, 960f)
                        } else tipLabel.draw(game.batch, 1f)
                    } catch (e: Exception) {}
                    game.batch.end()
                } else {
                    stage.draw()
                    labelStage.viewport.apply()
                    labelStage.draw()

                    //Special shape rendering here so it won't be blocked by labels
                    requestFlasher.update()
                }

                //Draw the UI overlay
                uiCam.update()
                if (finishedLoading) {
                    game.batch.projectionMatrix = uiCam.combined
                    uiStage.act()
                    uiStage.viewport.apply()
                    try {
                        uiStage.draw()
                    } catch (e: Exception) {}
                }
            }
            if (tutorialQuit) {
                TerminalControl.radarScreen = null
                game.screen = MainMenuScreen(game, null)
            }
        } catch (e: Exception) {
            ErrorHandler.sendGenericError(e, true)
        }
    }

    /** Check if tutorial is in paused state  */
    private fun checkTutorialPaused(): Boolean {
        return (this as RadarScreen).tutorialManager?.isPausedForReading ?: false
    }

    /** Implements resize method of screen, adjusts camera & viewport properties after resize for better UI  */
    override fun resize(width: Int, height: Int) {
        TerminalControl.WIDTH = width
        TerminalControl.HEIGHT = height
        ui.updatePaneWidth()
        viewport.update(width, height, false)
        stage.viewport.update(width, height, false)
        labelStage.viewport.update(width, height, false)
        val xOffset = camera.zoom * 990
        camera.position[camera.viewportWidth / 2 - xOffset, camera.viewportHeight / 2] = 0f
        uiViewport.update(width, height, true)
        uiStage.viewport.update(width, height, true)
        uiCam.position[uiCam.viewportWidth / 2f, uiCam.viewportHeight / 2f] = 0f
        if (Gdx.app.type == Application.ApplicationType.Desktop) {
            var resizeAgain = false
            var newWidth = width
            var newHeight = height
            if (newWidth < 960) {
                newWidth = 960
                resizeAgain = true
            }
            if (newHeight < 540) {
                newHeight = 540
                resizeAgain = true
            }
            if (resizeAgain) {
                resize(newWidth, newHeight)
            }
        }
    }

    /** Implements pause method of screen  */
    override fun pause() {
        setGameRunning(false)
    }

    /** Implements resume method of screen  */
    override fun resume() {
        setGameRunning(true)
    }

    /** Implements hide method of screen  */
    override fun hide() {
        //Not disposing unless game is quit specifically
    }

    /** Implements dispose method of screen, disposes resources after they're no longer needed  */
    override fun dispose() {
        stage.clear()
        stage.dispose()
        labelStage.clear()
        labelStage.dispose()
        uiStage.clear()
        uiStage.dispose()
        shapeRenderer.dispose()
        aircrafts.clear()
        airports.clear()
        waypoints.clear()
        soundManager.dispose()
    }

    /** Implements touchdown method of gestureListener  */
    override fun touchDown(x: Float, y: Float, pointer: Int, button: Int): Boolean {
        return false
    }

    /** Implements tap method of gestureListener, tests for tap and double tap  */
    override fun tap(x: Float, y: Float, count: Int, button: Int): Boolean {
        TerminalControl.radarScreen?.setSelectedAircraft(null)
        if (count == 2 && finishedLoading) {
            //Gdx.app.postRunnable{ obsArray = FileLoader.loadObstacles()} //Reload obstacles - Debug use only
            zooming = true
            return true
        }
        //Gdx.app.log("Coordinates", "${getWorldCoordFromScreenCoord(x, y)}") //For debug use
        return false
    }

    /** Calculates and returns position of mouse click in game world */
    private fun getWorldCoordFromScreenCoord(x: Float, y: Float): Vector2 {
        val screenHeightRatio = 3240f / Gdx.graphics.height
        val screenWidthRatio = 5760f / Gdx.graphics.width
        val vector3 = Vector3(x, y, 0f)
        val vector3New = stage.camera.unproject(vector3, 0f, 0f, Gdx.graphics.height * 5760f / 3240f, Gdx.graphics.height.toFloat())
        vector3New.x -= camera.zoom * (screenHeightRatio - screenWidthRatio) * Gdx.graphics.width / 2
        return Vector2(vector3New.x, vector3New.y)
    }

    /** Implements longPress method of gestureListener  */
    override fun longPress(x: Float, y: Float): Boolean {
        return false
    }

    /** Implements fling method of gestureListener  */
    override fun fling(velocityX: Float, velocityY: Float, button: Int): Boolean {
        return false
    }

    /** Implements pan method of gestureListener, tests for panning to shift radar screen around  */
    override fun pan(x: Float, y: Float, deltaX: Float, deltaY: Float): Boolean {
        if (!finishedLoading || dragging) {
            return false
        }
        val screenHeightRatio = 3240f / Gdx.graphics.height
        val screenWidthRatio = 5760f / Gdx.graphics.width
        val screenRatio = screenHeightRatio.coerceAtLeast(screenWidthRatio)
        camera.translate(-deltaX * camera.zoom * screenRatio, deltaY * camera.zoom * screenRatio)
        return false
    }

    /** Implements panStop method of gestureListener  */
    override fun panStop(x: Float, y: Float, pointer: Int, button: Int): Boolean {
        return false
    }

    /** Implements zoom method of gestureListener, tests for pinch zooming to adjust radar screen zoom level  */
    override fun zoom(initialDistance: Float, distance: Float): Boolean {
        if (!finishedLoading || distMode) {
            return true
        }
        if (initialDistance != lastInitialDist) {
            //New zoom
            lastInitialDist = initialDistance
            lastScale = camera.zoom
        }
        val ratio = lastInitialDist / distance
        camera.zoom = lastScale * ratio
        camera.update()
        return true
    }

    /** Implements pinch method of gestureListener  */
    override fun pinch(initialPointer1: Vector2, initialPointer2: Vector2, pointer1: Vector2, pointer2: Vector2): Boolean {
        return false
    }

    /** Implements pinchStop method of gestureListener  */
    override fun pinchStop() {
        //No default implementation
    }

    /** Implements keyDown method of inputListener  */
    override fun keyDown(keycode: Int): Boolean {
        return false
    }

    /** Implements keyUp method of inputListener  */
    override fun keyUp(keycode: Int): Boolean {
        return false
    }

    /** Implements keyTyped method of inputListener  */
    override fun keyTyped(character: Char): Boolean {
        return false
    }

    /** Implements touchdown method of inputListener  */
    override fun touchDown(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
        return false
    }

    /** Implements touchUp method of inputListener  */
    override fun touchUp(screenX: Int, screenY: Int, pointer: Int, button: Int): Boolean {
        return false
    }

    /** Implements touchDragged method of inputListener  */
    override fun touchDragged(screenX: Int, screenY: Int, pointer: Int): Boolean {
        return false
    }

    /** Implements mouseMoved method of inputListener  */
    override fun mouseMoved(screenX: Int, screenY: Int): Boolean {
        return false
    }

    /** Implements scrolled method of inputListener, tests for scrolling to adjust radar screen zoom levels  */
    override fun scrolled(amountX: Float, amountY: Float): Boolean {
        if (finishedLoading) {
            camera.zoom += amountY * 0.042f
        }
        return true
    }

    /** Sets whether game is running  */
    fun setGameRunning(running: Boolean) {
        this.running = running
        if (this.running) {
            val prevCamX = camera.position.x
            val prevCamY = camera.position.y
            if (game.screen !== this) game.screen = this
            Gdx.input.inputProcessor = inputMultiplexer
            camera.position.x = prevCamX
            camera.position.y = prevCamY
            soundManager.resume()
            DataTag.startTimers()
            DataTag.setBorderBackground()
        } else {
            soundManager.pause()
            DataTag.pauseTimers()
            game.screen = PauseScreen(game, this as RadarScreen)
        }
    }

    val finishedLoading
        get() = !metarLoading && uiLoaded
}